### ISSUE TITLE
Feature Request: Time-zone aware utilities

### ISSUE THREAD

**Issue by Griatch** [evennia/evennia#819]
#### Description of the suggested feature and how it is supposed to work for the admin/end user:

1. Store a property method on DefaultPlayer that allows for time-zone aware time conversion. This would both spit out current local time as a datetime+timezone object as well as convert a datetime+timezone object into local time. 
2. Use an on-Player Attrubute to store an optional custom timezone. 
3. Make a default Command for setting and retrieving this information.
4. Examine if it is possible to make an inlinefunc for inserting time in a locale-aware manner that is automatically translated to the right time for all viewers. 

#### A list of arguments for why you think this new feature should be included in Evennia:

1. Time determination across time zones is important for planning events/scenes/meetings in a range of different game styles.

#### Extra information, such as requirements or ideas on implementation:

It might not be possible to accomplish an inlinefunc that works like this; at least not while letting the string creator use their own local time. This is because the inlinefunc is only parsed once, for each session.

**Comment by Russell-Jones** [comment]
[Time zones in Django](https://docs.djangoproject.com/en/1.8/topics/i18n/timezones/) looks relevant.
0. lots of advice on using pytz
1. a check against pytz.common_timezones when setting in the player session
2. calling django.utils.timezone.activate(pytz.timezone(TZ)) looked handy, but seems to mainly be for templates and forms, which are not really used by Evennia.

Could the inlinefunc use kwargs.get("session").tz to convert to UTC with some kind of marker indicating it's a time, then code elsewhere (player.msg?) convert to the viewers' timezone?

**Comment by Griatch** [comment]
@Russell-Jones 

Yes, I think handling the timezone itself is well cared for by Django as you point out (also Python itself has some nice features on this). 

The problem with the inlinefunc is that it doesn't have a "state" in itself - it's literally a tag in the text that Evennia parses for every new session that receives text with that marker. 

I suppose one way to do this would be to let the creator himself specify their timezone as part of the call signature to the inlinefunc, either by an identifier or by the time offset. Not quite as automatic as I would like but it would be something like

```
say Let's have that meet at {mytime(14:00, EST)!
say The next quest starts at {mytime(15:15, -1)
```
If marked this way, the system could easily translate it to the local time for every player seeing the string. Maybe it should it say which time zone it's using, just to avoid the inavoidable questions.

So I would see

```
Dude says, "Let's have that meet at 21:00(CET)!"
Dude says, "The next quest starts at 17:15(CET)"
```

Something like that.

**Comment by Griatch** [comment]
This can be implemented using nested inlinefuncs `$mytime(14:00, EST)`.

**Comment by strikaco** [comment]
There is a problem with that in that acronyms can be ambiguous if you have an international userbase. `CST` can mean China Standard Time, Cuba Standard Time or [American] Central Standard Time. If we settle on America being the One True Source of Time, then sure, it works as expected :)

The proper way to do it would be to spell out the zone, i.e. `America/New_York`, or use numeric GMT offsets which are less intuitive.

**Comment by Griatch** [comment]
@strikaco These would be the time zones we'd support I guess: 

```python
import pytz
for tz in pytz.all_timezones:
    print tz
```
One just need to make them clear (there is no CST there as far as I can see)

**Comment by chainsol** [comment]
Hmmm, would this fit better as some nature of contrib or something?

I think there's pretty "easy" ways to do this, but it feels out-of-place in the inlinefuncs code, especially as it won't be something a player can just manage to do correctly, since you do need to use the technically correct timezone name, or an offset.

I've got a little inlinefunc written, but invalid inputs like $mytime(00:61, Pickletime) can't really raise an error or anything, they have to just output something like "invalid time" or something.

**Comment by Griatch** [comment]
@strikaco This issue is actually only tangentially about the inlinefunc and more about the time-aware utilities that would surround it ;)

**Comment by Griatch** [comment]
(see the original issue description)

**Comment by chainsol** [comment]
Noted - I do think the inlinefunc is one of the harder things due to the messiness of the input. Would two methods on the DefaultAccount be acceptable - one to convert datetime+tz objects to local time, and one to get local time?

**Comment by strikaco** [comment]
@Griatch That method still returns some ambiguities-- there is in fact CST6CDT, which is Central Standard Time...for Americans. Same with EST and PST. `pytz.common_timezones` as @Russell-Jones mentioned would be safer; it does away with most of the acronyms altogether.

Here's a related discussion about the complexities of timezone handling:
https://news.ycombinator.com/item?id=10032498

@chainsol 
> I forgot to mention, this facility would also require each account to store the local time zone for said account, so this would need to be added too. It could default to the server's time.

That's a logical attribute to implement; most web applications handling user accounts will ask for or let you set timezone information for this very reason. 

The server default timezone is UTC, but only computers and military personnel think in terms of UTC. Civilians always think in terms of their own (source) timezone or the timezone of an event's location (destination). If someone tells you nothing more than "let's meet at the Bellagio at 9" and vanishes, one could debate whether they meant 9 in the current local timezone (if outside Nevada) or 9 in Pacific Time, but under no circumstance should we assume they meant 9:00 UTC+0 unless your game is so heavily military-themed even event coordination must be done in character :)

So without a tz attribute set and without geoip, we don't know where the user is and can't reasonably assume what timezone they meant. Interpreting that timestamp should fail IMO.

**Comment by Griatch** [comment]
@chainsol Not sure about it; Python/pytz has all resources itself to convert so a dev can do that too if needed. I would probably say that an Account method should always return the user's time expressed in the Server's set time (UTC by default), but that there would a keyword arg to have it return user's local time instead? 

@strikaco 
I live at UTC+1, I would much prefer if everyone used UTC to announce events, would make it a lot easier to count for everyone around the world ;) Coming to think of it, it may be a good idea to show the server's set time zone with the `@about` command ...

Anyway, I don't think there is anyone with any internet (or MU) experience who are not aware of the fact that people live at different time zones. You are definitely aware of it you know to use the inlinefunc in the first place. An inlinefunc apply to every account, including the one creating the string in which it resides. So the player will get back string with the time in it replaced too. The time they see should match the time they expected in their local time. If it does not it means they should either supply it explicitly to the function or change whatever setting they have.
